### PR TITLE
feat(grpo-eval): add GRPO evaluation component

### PR DESCRIPTION
**Pull request by Sridhar1030**
## Summary

- Adds a lightweight, CPU-only `grpo_eval` Kubeflow Pipelines component.
- Reads `training_results.json` from a caller-provided shared-PVC path.
- Logs GRPO metrics: `mean_reward`, `full_match_rate`, initial/final reward, reward improvement, and training iterations.
- Returns a named boolean output: `promotion_passed`.
- Adds metadata, ownership, generated documentation, package registration, and unit tests.

## Contract

The JSON/PVC layout is currently provisional until the `lora_grpo` training component confirms its final output contract. The component accepts the full results-file path, so future path/layout changes do not require changing its public API.

## Testing

- `uv run --extra dev pytest components/evaluation/grpo_eval/tests/test_component_unit.py -v`
- Ruff formatting and lint checks
- KFP component compilation
- Metadata, README, package-entry, base-image, example, and import validations
- OpenShift KFP smoke test passed:
  - Shared workspace PVC provisioned successfully
  - `training_results.json` read from `/kfp-workspace/grpo/training_results.json`
  - `promotion_passed=true`
  - Expected metrics recorded

Ref: https://redhat.atlassian.net/browse/RHOAIENG-90695

## Summary by CodeRabbit

- **New Features**
  - Added an alpha GRPO evaluation component that validates shared-workspace training results, reports reward, match-rate, and timing metrics, and indicates whether results meet promotion criteria.
  - Added a self-contained reward chart showing mean reward by training iteration.

- **Documentation**
  - Added guidance covering inputs, validation rules, outputs, metadata, dependencies, and availability.

- **Bug Fixes**
  - Added safeguards for extreme reward ranges that cannot be rendered safely in charts.

- **Tests**
  - Added coverage for successful evaluations, promotion decisions, chart and metric outputs, compilation, and invalid or missing inputs.

### CODE REVIEW

**Comment by Sridhar1030** [comment]
Ran a demo pipline results :-
image
image

**Comment by efazal** [comment]
Thanks @Sridhar1030, the component looks well structured. Left a few comments to address.

File: components/evaluation/grpo_eval/component.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+"""GRPO evaluation component."""
+
+from typing import NamedTuple
+
+from kfp import dsl
+
+
+@dsl.component(base_image="python:3.11")
```
Comment (efazal) [inline components/evaluation/grpo_eval/component.py]: The sibling `lm_eval` component uses `registry.access.redhat.com/ubi9/python-311:latest` — Docker Hub images won't pass image scanning for the Red Hat distribution. Please switch to the UBI image (or whatever pinned digest the other components in this repo use).

File: components/evaluation/grpo_eval/component.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    normalized_timings = [
+        require_finite_number(timing, f"timing_history[{index}]") for index, timing in enumerate(timing_history)
+    ]
+    initial_reward = normalized_rewards[0]
+    final_reward = normalized_rewards[-1]
+    full_match_rate = normalized_match_rates[-1]
+    reward_improvement = final_reward - initial_reward
+    promotion_passed = len(normalized_rewards) > 1 and final_reward > initial_reward
+
+    output_metrics.log_metric("mean_reward", mean_reward)
```
Comment (efazal) [inline components/evaluation/grpo_eval/component.py:178]: nit: `final_mean_reward` from JSON is logged here as `mean_reward`, while `reward_history[-1]` is logged as `final_reward` on line 107. In the current fixture these are the same value, but if ART ever computes them differently the KFP UI will show two metrics that both claim to be the final reward. Worth either consolidating or documenting what each represents.

File: components/evaluation/grpo_eval/component.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    output_metrics.log_metric("initial_reward", initial_reward)
+    output_metrics.log_metric("final_reward", final_reward)
+    output_metrics.log_metric("reward_improvement", reward_improvement)
+    output_metrics.log_metric("training_iterations", float(len(normalized_rewards)))
+    if normalized_timings:
+        output_metrics.log_metric("initial_iteration_time_seconds", normalized_timings[0])
+        output_metrics.log_metric("final_iteration_time_seconds", normalized_timings[-1])
+        output_metrics.log_metric("mean_iteration_time_seconds", sum(normalized_timings) / len(normalized_timings))
+
+    return NamedTuple("GrpoEvalOutputs", [("promotion_passed", bool)])(
```
Comment (efazal) [inline components/evaluation/grpo_eval/component.py:190]: nit: `promotion_passed` drives the downstream gate but isn't logged to `output_metrics`, so it won't appear in the KFP run UI metrics panel. Adding `output_metrics.log_metric("promotion_passed", float(promotion_passed))` would make the decision visible alongside the reward numbers.

**Comment by Sridhar1030** [comment]
@ChughShilpa Added reward curve 
 
image

File: components/evaluation/grpo_eval/component.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
+"""GRPO evaluation component."""
+
+from typing import NamedTuple
+
+from kfp import dsl
+
+
+@dsl.component(base_image="registry.access.redhat.com/ubi9/python-311:latest")
```
Comment (ChughShilpa) [inline components/evaluation/grpo_eval/component.py]: Use this as base image `quay.io/opendatahub/odh-th-torch-cpu-py312:odh-3.6-ea.2`

Comment (ChughShilpa) [reply components/evaluation/grpo_eval/component.py]: Wondering this odh torch-cpu image includes PyTorch, which is heavy for a CPU-only eval component that only uses json and math from stdlib. 
Let's strict to `registry.access.redhat.com/ubi9/python-311:latest` as other eval components are using it @Sridhar1030

Comment (Sridhar1030) [reply components/evaluation/grpo_eval/component.py]: Oh ok , ill revert the commit back to the UBI Python image and keep the KFP runtime installation explicit.

File: components/evaluation/grpo_eval/component.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            raise ValueError(f"'{field_name}' must be a finite number")
+        return number
+
+    def write_reward_chart(rewards: list[float], path: str) -> None:
+        """Write a self-contained SVG reward curve as a KFP HTML artifact."""
+        chart_width, chart_height, margin = 800, 400, 70
+        plot_width = chart_width - 2 * margin
+        plot_height = chart_height - 2 * margin
+        reward_min, reward_max = min(rewards), max(rewards)
+        reward_range = reward_max - reward_min
```
Comment (ChughShilpa) [inline components/evaluation/grpo_eval/component.py:67]: This `reward_range = reward_max - reward_min` overflows to inf when rewards span extreme finite ranges (e.g. [-sys.float_info.max, sys.float_info.max]). Downstream SVG coordinates become NaN/inf, producing an invalid chart.

You can clamp or normalize reward values before coordinate calculation, or reject reward ranges where reward_max - reward_min is not finite. 
Also suggestion to add a regression test with extreme-value rewards.

File: components/evaluation/grpo_eval/tests/test_component_unit.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    expected_promotion: bool,
+):
+    """Only strict, multi-iteration reward improvement passes promotion."""
+    results_path = write_results(
+        tmp_path / "training_results.json",
+        {
+            "final_mean_reward": 0.5,
+            "reward_history": reward_history,
+            "full_match_history": [0.5] * len(reward_history),
+            "timing_history": [],
```
Comment (ChughShilpa) [inline components/evaluation/grpo_eval/tests/test_component_unit.py:135]: timing_history is allowed to be empty while full_match_history must be non-empty and match reward_history length. 
The empty-timing path is tested (non-improving promotion tests use timing_history: []), but no test verifies that timing metrics are omitted from the logged output when empty. Worth adding a targeted assertion.

File: components/evaluation/grpo_eval/tests/fixtures/training_results.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+  "final_mean_reward": 0.67,
+  "reward_history": [0.33, 0.48, 0.67],
```
Comment (ChughShilpa) [inline components/evaluation/grpo_eval/tests/fixtures/training_results.json:3]: I see you added a docstring ("The logged mean_reward is ART's reported final aggregate reward, while final_reward is the last entry in reward_history...").
 In the test fixture these happen to be the same value (0.67), but if ART ever computes them differently the KFP UI will show two metrics that both look like "the final reward."

Just a suggestion not a hard requirement: So consider a test fixture where these values diverge to make the distinction explicit.

**Review by ChughShilpa — APPROVED** [review]

**Review by efazal — APPROVED** [review]
Thanks @Sridhar1030!
/lgtm

**Comment by efazal** [comment]
@mprahl @HumairAK @hbelmiro can we get your approval on this PR please as we don't have permission to merge under the evaluation components. This is for 3.6 EA2, would really appreciate your help here. Thanks!

**Review by mprahl — APPROVED** [review]
/approve
/lgtm

Approving to unblock the PR since the owners have reviewed it.
